### PR TITLE
feat: new limited jurisdictional admin role

### PR DESCRIPTION
**Pull request by mcgarrye**
This PR addresses [#(490)](https://app.zenhub.com/workspaces/bloom-5dc32d7144bd400001315dac/issues/gh/metrotranscom/doorway/490)

- [x] Addresses the issue in full
- [ ] Addresses only certain aspects of the issue

## Description
New limited jurisdictional admin role for Doorway. This role should have all the same permissions as an standard jurisdictional admin, but without access to any PII. This means they CANNOT do the following: see the partners users tab, see/edit/delete partner or public users, see/edit/delete listing applications, submit paper applications, see/resolve/process application flagged sets (aka application duplicates). This should be across the partners site and the API.

## How Can This Be Tested/Reviewed?

Test a user with the new role can be created. To do so, login as an admin user to the partners portal, navigate to the users tab, add a new user and select Jurisdictional Admin - No PII.

Use seeded or newly created limited jurisdictional admin to test all listing and settings flows successfully. Check that the user cannot do anything of the actions outlined in the description.

## Author Checklist:

- [x] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [ ] Reviewed in a mobile view
- [ ] Reviewed considering accessibility
- [x] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [x] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

### CODE REVIEW

**Review by ludtkemorgan — CHANGES_REQUESTED** [review]
When I create an account with this new role type using my email the MFA code always returns an error 
image
and if I submit it again I get this error 
image.

This does not happen when creating an admin account so I'm guessing adding this new role was just missed somewhere. Other than that mostly minor comments

File: api/src/services/application-csv-export.service.ts [hunk 1]
```diff
@@ -871,6 +871,9 @@ export class ApplicationCsvExporterService
      * By making listingId required, we can check if the user has update permissions for the listing, since right now if a user has that
      * they also can run the export for that listing
      */
+    if (user?.userRoles?.isLimitedJurisdictionalAdmin)
```
[the remark this answers is not in the document]
Comment (mcgarrye) [reply api/src/services/application-csv-export.service.ts:874]: The difficulty around using casbin is that on line 880 we are calling the permissions with `listing` as the type, which this user has access to, but this is really application data so adding this check was the simplest way to check without completely changing the existing check

File: sites/partners/page_content/locale_overrides/general.json [hunk 2]
```diff
@@ -485,6 +485,7 @@
   "users.editUser": "Edit User",
   "users.jurisdictionalAdmin": "Jurisdictional Admin",
   "users.jurisdictionalizedListings": "%{jurisdiction} Listings",
+  "users.limitedJurisdictionalAdmin": "Jurisdictional Admin - No PII",
```
[the remark this answers is not in the document]
Comment (mcgarrye) [reply sites/partners/page_content/locale_overrides/general.json:488]: I can mention it to Sarah

**Comment by mcgarrye** [comment]
> When I create an account with this new role type using my email the MFA code always returns an error image and if I submit it again I get this error image.
> 
> This does not happen when creating an admin account so I'm guessing adding this new role was just missed somewhere. Other than that mostly minor comments

I did see this once. You should be able to get in after using the mfa code and then going to localhost:3001 without anything else in the url. Will test again and fix

File: sites/partners/page_content/locale_overrides/general.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcgarrye) [reply sites/partners/page_content/locale_overrides/general.json:488]: Confirmed that this is okay

File: api/src/guards/admin-no-pii.guard.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
+import { User } from '../dtos/users/user.dto';
+
+@Injectable()
+export class AdminNoPIIGuard implements CanActivate {
```
Comment (emilyjablonski) [inline api/src/guards/admin-no-pii.guard.ts:5]: I'm finding the naming of this to be confusing - it's a guard for no PII, but does allow limited jurisdictional admins?

File: api/src/permission-configs/permission_policy.csv [hunk 4]
```diff
@@ -11,10 +11,12 @@ p, anonymous, user, true, create
...
 p, jurisdictionAdmin, asset, true, .*
+p, limitedJurisdictionAdmin, asset, true, .*
 p, partner, asset, true, .*
 
 p, admin, multiselectQuestion, true, .*
 p, jurisdictionAdmin, multiselectQuestion, true, .*
+p, limitedJurisdictionAdmin, multiselectQuestion, true, .*
 p, partner, multiselectQuestion, true, .*
 
 p, admin, applicationMethod, true, .*
```
Comment (emilyjablonski) [inline api/src/permission-configs/permission_policy.csv:22]: I think this should be true for the new role

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,1265 @@
...
+      });
+
+      await request(app.getHttpServer())
+        .get(`/applications/${applicationA.id}`)
+        .set({ passkey: process.env.API_PASS_KEY || '' })
+        .set('Cookie', cookies)
+        .expect(403);
+    });
+
+    it('should error as forbidden for delete endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:253]: Looks like we can delete `& create an activity log entry`

File: api/src/permission-configs/permission_policy.csv [hunk 6]
```diff
@@ -49,25 +56,30 @@ p, admin, translation, true, .*
...
 p, jurisdictionAdmin, unitRentType, true, read
+p, limitedJurisdictionAdmin, unitRentType, true, read
 p, partner, unitRentType, true, read
 
 p, admin, unitAccessibilityPriorityType, true, .*
 p, jurisdictionAdmin, unitAccessibilityPriorityType, true, read
+p, limitedJurisdictionAdmin, unitAccessibilityPriorityType, true, read
 p, partner, unitAccessibilityPriorityType, true, read
 
 p, admin, applicationMethod, true, .*
```
[the remark this answers is not in the document]
Comment (mcgarrye) [reply api/src/permission-configs/permission_policy.csv:85]: It is duplicated in the file, I created a ticket to fix it in core and elsewhere. Didn't want to broaden the scope of the changes in an already large PR

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,1265 @@
...
+        .delete(`/applications/`)
+        .set({ passkey: process.env.API_PASS_KEY || '' })
+        .send({
+          id: applicationA.id,
+        })
+        .set('Cookie', cookies)
+        .expect(403);
+    });
+
+    it('should succeed for public create endpoint', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:282]: The new role should be able to create an application?

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,1265 @@
...
+            listing1Created.id,
+            unitTypeA.id,
+            new Date(),
+          ),
+        )
+        .set('Cookie', cookies)
+        .expect(201);
+    });
+
+    it('should error as forbidden for partner create endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:309]: Looks like we can delete `& create an activity log entry`

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,1265 @@
...
+            listing1Created.id,
+            unitTypeA.id,
+            new Date(),
+          ),
+        )
+        .set('Cookie', cookies)
+        .expect(403);
+    });
+
+    it('should error as forbidden for update endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:336]: Looks like we can delete `& create an activity log entry`

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts (hunk 7 above)
Comment (mcgarrye) [reply api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:282]: This is a public application. Anyone can create an application on the public site

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,1265 @@
...
+            listing1Created.id,
+            unitTypeA.id,
+            new Date(),
+          ),
+        )
+        .set('Cookie', cookies)
+        .expect(201);
+    });
+
+    it('should error as forbidden for csv endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:399]: Looks like we can delete `& create an activity log entry`

File: api/src/services/listing-csv-export.service.ts [hunk 11]
```diff
@@ -851,6 +851,7 @@ export class ListingCsvExporterService implements CsvExporterServiceInterface {
       user &&
       (user.userRoles?.isAdmin ||
         user.userRoles?.isJurisdictionalAdmin ||
+        user.userRoles?.isLimitedJurisdictionalAdmin ||
```
Comment (emilyjablonski) [inline api/src/services/listing-csv-export.service.ts:854]: Is the new role meant to be able to export the CSV? We have a new test `should error as forbidden for csv endpoint`

Comment (mcgarrye) [reply api/src/services/listing-csv-export.service.ts:854]: The new role should be able to export the listing csv

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,1265 @@
...
+        .send(
+          // builds an invite for an admin
+          buildUserInviteMock(jurisId, 'partnerUser+jurisCorrect@email.com'),
+        )
+        .set('Cookie', cookies)
+        .set({ passkey: process.env.API_PASS_KEY || '' })
+        .expect(403);
+    });
+
+    it('should error as forbidden for csv export endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-correct-juris.e2e-spec.ts:976]: Looks like we can delete `& create an activity log entry`

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,1235 @@
...
+  confirmation: jest.fn(),
+  welcome: jest.fn(),
+  invitePartnerUser: jest.fn(),
+  changeEmail: jest.fn(),
+  forgotPassword: jest.fn(),
+  sendMfaCode: jest.fn(),
+  applicationConfirmation: jest.fn(),
+};
+
+describe('Testing Permissioning of endpoints as Jurisdictional Admin in the wrong jurisdiction', () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts:77]: `as Limited Jurisdictional Admin`

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,1235 @@
...
+      });
+
+      await request(app.getHttpServer())
+        .get(`/applications/${applicationA.id}`)
+        .set({ passkey: process.env.API_PASS_KEY || '' })
+        .set('Cookie', cookies)
+        .expect(403);
+    });
+
+    it('should error as forbidden for delete endpoint & create an activity log entry', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts:256]: Looks like we can delete `& create an activity log entry`

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,1235 @@
...
+        .delete(`/applications/`)
+        .set({ passkey: process.env.API_PASS_KEY || '' })
+        .send({
+          id: applicationA.id,
+        })
+        .set('Cookie', cookies)
+        .expect(403);
+    });
+
+    it('should succeed for public create endpoint', async () => {
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts:285]: Same q in terms of application creation

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,1235 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (emilyjablonski) [inline api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts:1]: At a high level, what is different in terms of permissions if the user is in the wrong jurisdiction?

File: api/test/unit/services/user-csv-export.service.spec.ts [hunk 17]
```diff
@@ -140,6 +140,11 @@ describe('Testing user csv export service', () => {
                     isJurisdictionalAdmin: true,
                   },
                 },
+                {
```
Comment (emilyjablonski) [inline api/test/unit/services/user-csv-export.service.spec.ts:143]: I didn't think the new role was supposed to be able to see users?

File: sites/partners/src/pages/listings/[id]/applications/add.tsx [hunk 18]
```diff
@@ -10,33 +11,35 @@ import { useSingleListingData } from "../../../../lib/hooks"
...
-            <BreadcrumbLink href={`/listings/${listingId}/applications/add`} current>
-              {t("t.add")}
-            </BreadcrumbLink>
-          </Breadcrumbs>
-        }
-      />
-
-      <PaperApplicationForm listingId={listingId} />
-    </Layout>
+    !profile?.userRoles?.isLimitedJurisdictionalAdmin && (
```
Comment (emilyjablonski) [inline sites/partners/src/pages/listings/[id]/applications/add.tsx:19]: Would prefer a check before the return, like if (!role) return null

Comment (mcgarrye) [reply sites/partners/src/pages/listings/[id]/applications/add.tsx:19]: I did attempt this but it threw errors about having multiple returns. could try again

File: api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts (hunk 16 above)
Comment (mcgarrye) [reply api/test/integration/permission-tests/permission-as-limited-juris-admin-wrong-juris.e2e-spec.ts:1]: This is the same for jurisdictionAdmins, limitedJurisdictionalAdmins, and partners. They should not be able to see/create/update/delete any listing in jurisdiction they don't have access to. And they shouldn't be able to see any applications or users associated with listings they don't have access to

File: api/test/unit/services/user-csv-export.service.spec.ts (hunk 17 above)
Comment (mcgarrye) [reply api/test/unit/services/user-csv-export.service.spec.ts:143]: This adds limited jurisdictional admins to the user list. It adds all users except for admin

File: api/src/permission-configs/permission_policy.csv (hunk 4 above)
Comment (mcgarrye) [reply api/src/permission-configs/permission_policy.csv:22]: I am able to add/change applicationMethod. There is no calls using canOrThrow with this type. Can probably be removed going forward.

File: sites/partners/src/pages/listings/[id]/applications/add.tsx (hunk 18 above)
Comment (mcgarrye) [reply sites/partners/src/pages/listings/[id]/applications/add.tsx]: Worked this time!

**Review by YazeedLoonat — APPROVED** [review]
LGTM

**Review by emilyjablonski — APPROVED** [review]
🚛

**Review by ludtkemorgan — APPROVED** [review]

**Pull request merged**